src/main/scala/mix/kafka/monitor/LibratoOffsetReporter.scala

```diff
@@ -1,86 +1,95 @@
 package mix.kafka.monitor
 
+import java.lang.{Integer => JInt, Long => JLong}
 import java.time.Duration
 import java.util.concurrent.TimeUnit
 
 import com.codahale.metrics.{Gauge, MetricRegistry, ScheduledReporter}
 import com.quantifind.kafka.OffsetGetter
-import com.quantifind.kafka.OffsetGetter.OffsetInfo
 import kafka.utils.Logging
 
-class LibratoOffsetReporter(metrics: MetricRegistry,
+class LibratoOffsetReporter(metricRegistry: MetricRegistry,
                             reporter: ScheduledReporter,
-                            reportingInterval: Duration,
-                            metricsCacheExpiration: Duration) extends Logging {
+                            cacheProvider: MetricsCacheProvider,
+                            source: String,
+                            reportingInterval: Duration) extends Logging {
 
   import LibratoOffsetReporter._
 
-  logger.info(s"starting LibratoOffsetReporter with interval = $reportingInterval, cacheTTL = $metricsCacheExpiration")
+  logger.info(s"starting LibratoOffsetReporter with source = $source, interval = $reportingInterval")
 
   reporter.start(reportingInterval.getSeconds, TimeUnit.SECONDS)
 
   def report(offsets: IndexedSeq[OffsetGetter.OffsetInfo]): Unit = {
-    val cacheEntries = offsets map { offset =>
-      getMetricName(offset) -> Lag(offset.lag)
-    } toMap
-
-    import scala.collection.JavaConverters._
-
-    /*
-     * First fetching all keys from cache before updating them since that would trigger registration of new metrics
-     * (via cacheLoader) to dropwizard-metrics registry without errors since dropwizard throw errors while trying to
-     * register existing gauges - so basically relying on cache synchronised behavior to register metrics only once.
-     */
-    cache.getAll(cacheEntries.keys.asJava)
-    cache.putAll(cacheEntries.asJava)
+    val metricsWithUpdatedValues = offsetsToMetrics(offsets)
+    val existingMetrics = cacheProvider.existingCacheEntries(metricsWithUpdatedValues.keys.toSeq)
+    val newMetrics = metricsWithUpdatedValues -- existingMetrics.keys
+    val mergedMetrics = mergeMetrics(existingMetrics, metricsWithUpdatedValues)
```

Comment:
let me say what I can do. The problem here is - dealing with Map of Map of java types. Including it everywhere makes it even more non-readable. I had it initially and removed it later but tried out following things to solve this concern.

I have created intermediate methods with clear names and explicit return types to try to make it manageable. Other minor thing I intentionally did was to keep the types same. So for instance, in the above method call - `mergeMetrics` both inputs and the return type have same types. Maybe I could include explicit type for the first one to make it better.